### PR TITLE
Add colossatr0n's email and storage improvements

### PR DESCRIPTION
**Pull request by jdanders**
When this fork stabilizes and is pep8 compliant it can be merged.

### CODE REVIEW

**Comment by colossatr0n** [comment]
I'll comment here again when this is ready to be merged. It should be code complete unless I find a bug while testing out the latest commit.

**Comment by colossatr0n** [comment]
This PR is now ready.

File: ksl.py [hunk 1]
```diff
@@ -133,21 +158,26 @@ def build_qs(self, queries, **etc):
...
+            logging.debug("Generated the search URL: {url}".format(url=queryurl))
             yield (query, queryurl, )
 
-    def listing(id):
+    def listing(self, id):
         pass
 
 
 def gather_report(query_result):
+    logging.debug("Gathering report for {count} listings".format(count=len(query_result)))
```
Comment (jdanders) [inline ksl.py:183]: `query_result` is a generator and has no `len`. If you really want a number here, you would need to create a list and then adjust the `for` loop below to use that list.

File: README.md [hunk 2]
```diff
@@ -62,18 +58,49 @@ optional arguments:
                         is newest to oldest
   -s, --sold            If included, query will return results for sold items
                         as well as active items
+  --email EMAIL         email address to send emails from.If --receiver is not
```
Comment (jdanders) [inline README.md]: Needs a space after the period.

File: README.md [hunk 3]
```diff
@@ -62,18 +58,49 @@ optional arguments:
...
+                        specified, this email will also be used as the
+                        receiver.
+  --receiver RECEIVER   email address to send the email to. Defaults to
+                        --email value.
+  --exception-receiver EXCEPTION_RECEIVER
```
Comment (jdanders) [inline README.md]: These param names work, but I would prefer `--to-email` and `--exception-email`. When reading just the list of parameters without help descriptions I had no idea what `receiver` might do.

File: README.md [hunk 4]
```diff
@@ -62,18 +58,49 @@ optional arguments:
...
+                        will be included.
+  --char-limit CHAR_LIMIT
+                        Number of characters allowed in the message body.
+                        Listings that exceed the character count will be sent
+                        in additional messages.
+  --exclude-links       Exclude links from message.
+  --load LOAD           Load seen listings from a JSON file. Format is a
+                        dictionary of query search terms to listing links.
+  --save SAVE           Save seen listings to a JSON file. File extension must
+                        be .json.
```
Comment (jdanders) [inline README.md]: This is not true. The file format will be json, but nothing is requiring the file extension be .json. I'd recommend changing the help to remove the sentence "File extension must be .json"

File: README.md [hunk 5]
```diff
@@ -62,18 +58,49 @@ optional arguments:
...
+  -l LOGFILE, --logfile LOGFILE
+                        File to log output from daemon process, defaults to
+                        stdout
+  --loglevel LOGLEVEL   Choose level: debug, info, warning
   -f, --foreground      Do not fork to background
+  -e EMAILEXCEPTIONS, --emailexceptions EMAILEXCEPTIONS
```
Comment (jdanders) [inline README.md]: `--email-exceptions` would match the style better.

All the above comments need changes in the arg parser itself in addition to the README, obviously.

File: README.md [hunk 6]
```diff
@@ -62,18 +58,49 @@ optional arguments:
...
+                        description. If not specified, the entire description
+                        will be included.
+  --char-limit CHAR_LIMIT
+                        Number of characters allowed in the message body.
+                        Listings that exceed the character count will be sent
+                        in additional messages.
+  --exclude-links       Exclude links from message.
+  --load LOAD           Load seen listings from a JSON file. Format is a
+                        dictionary of query search terms to listing links.
+  --save SAVE           Save seen listings to a JSON file. File extension must
```
Comment (jdanders) [inline README.md]: These all could benefit from short options as well. I'd recommend:
* `-h` `--head`
* `-H` `--char-limit`
* `-x` `--exclude-links`
* `-L` `--load`
* `-S` `--save`

File: README.md [hunk 7]
```diff
@@ -62,18 +58,49 @@ optional arguments:
...
+                        Number of repeated exceptions before sending emails
 ```
 
 ### Advanced Usage:
 
 Most filters available on the KSL webpage are made available through the script. Here's an example:
 
 ```
-./ksl_notify.py iphone galaxy --email example@example.com --smtpserver "smtp.example.com:587" -l /tmp/ksl_iphone_log.log --foreground --category Electronics --subcategory "Cell Phones Unlocked" --min-price 100 --max-price 250 --zip 84111 --miles 35 --time 60
+./ksl_notify.py iphone galaxy --sender sender@example.com --receiver receiver@example.com --smtpserver "smtp.example.com:587" -l /tmp/ksl_iphone_log.log --foreground --category Electronics --subcategory "Cell Phones Unlocked" --min-price 100 --max-price 250 --zip 84111 --miles 35 --time 60
```
Comment (jdanders) [inline README.md]: There is no `--sender` parameter. Keeping it named `--email` and defaulting that to both sender and receiver is good, so just change this description back.

File: ksl.py [hunk 8]
```diff
@@ -31,7 +46,9 @@ class KSL(object):
     def __init__(self):
         self.thread_pool = concurrent.futures.ThreadPoolExecutor(max_workers=4)
 
-    def __do_request(self, args):
+    @staticmethod
```
Comment (jdanders) [inline ksl.py:49]: I don't see a reason for this to be decorated a `staticmethod`, it's not called outside the class.

**Review by jdanders — COMMENTED** [review]
Thanks so much for the contribution! There's one actual error in `ksl.py` line 180, but apart from that I just have some stylistic comments.

File: ksl_notify.py [hunk 9]
```diff
@@ -133,43 +152,57 @@ def main(args):
         logging.getLogger().setLevel(numeric_level)
 
     # Get needed controls
-    loop_delay = args.pop('time') * 60
-    email = args.pop('email', None)
+    minutes = args.pop('time')
+    loop_delay = minutes * 60
+    sender = args.pop('email', None)
```
Comment (jdanders) [inline ksl_notify.py:162]: Please test the `load` parameter if present and valid before prompting for email. If the user has messed up the parameter, it's best to report the problem before asking for input.

File: ksl.py (hunk 8 above)
Comment (colossatr0n) [reply ksl.py:49]: A method being static or non-static doesn't have any bearing on whether or not it's called externally. A method being static or non-static depends on whether or not it accesses or mutates the state of object. If a class method doesn't rely on any object state (`self`), then it should be static.

**Comment by colossatr0n** [comment]
Sorry it took me so long to get back to this. Just pushed the changes.

File: ksl.py (hunk 1 above)
Comment (colossatr0n) [reply ksl.py:183]: Done.

File: README.md (hunk 2 above)
Comment (colossatr0n) [reply README.md]: Done.

File: README.md (hunk 4 above)
Comment (colossatr0n) [reply README.md]: Done.

File: README.md (hunk 5 above)
Comment (colossatr0n) [reply README.md]: Done.

File: README.md (hunk 6 above)
Comment (colossatr0n) [reply README.md]: Done. Had to change some of the short options due to conflicting option names.

File: README.md (hunk 7 above)
Comment (colossatr0n) [reply README.md]: Done.

File: ksl_notify.py (hunk 9 above)
Comment (colossatr0n) [reply ksl_notify.py:162]: Done.

**Pull request merged**